### PR TITLE
CBG-608: Avoid sprintf in bucket op retry descriptions

### PR DESCRIPTION
**Pull request by JRascagneres**
- [x] Awaiting Hydrogen dev branch
- [ ] http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration/50/
- [ ] http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration/53/

Couple of failing tests in above but the failing tests seem to run fine locally, I'm guessing its flaky tests. Nothing looks like it'd have the cause of this change.

### CODE REVIEW

File: base/bucket_gocb.go [hunk 1]
```diff
@@ -353,8 +353,12 @@ func (bucket *CouchbaseBucketGoCB) Get(k string, rv interface{}) (cas uint64, er
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("Get %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("Get", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		if _, ok := err.(*RetryTimeoutError); ok {
```
Comment (bbrks) [inline base/bucket_gocb.go]: How come we're checking for only this error now? Shouldn't it be kept as all errors?

File: base/bucket_gocb.go [hunk 2]
```diff
@@ -1421,8 +1443,13 @@ func (bucket *CouchbaseBucketGoCB) GetXattr(k string, xattrKey string, xv interf
 		}
 
 	}
-	description := fmt.Sprintf("GetXattr %s", UD(k).Redact())
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+
+	err, result := RetryLoop("GetXattr", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		if _, ok := err.(*RetryTimeoutError); ok {
```
Comment (bbrks) [inline base/bucket_gocb.go]: Also here - I think we still need to handle all types of error. Not just timeout.

File: base/bucket_n1ql.go [hunk 3]
```diff
@@ -248,10 +248,12 @@ func (bucket *CouchbaseBucketGoCB) WaitForIndexOnline(indexName string) error {
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("WaitForIndexOnline for index %s", indexName)
-	err, _ := RetryLoop(description, worker, CreateMaxDoublingSleeperFunc(25, 100, 15000))
+	err, _ := RetryLoop("WaitForIndexOnline", worker, CreateMaxDoublingSleeperFunc(25, 100, 15000))
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "WaitForIndexOnline for index %s", indexName)
```
Comment (bbrks) [inline base/bucket_n1ql.go]: We need to return here now if we're switching the return below.

File: base/bucket_gocb.go [hunk 4]
```diff
@@ -2479,8 +2507,10 @@ func (bucket *CouchbaseBucketGoCB) GetExpiry(k string) (expiry uint32, getMetaEr
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("getExpiry for key: %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("GetExpiry", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "Error during GetExpiry for key: %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 5]
```diff
@@ -1818,8 +1845,10 @@ func (bucket *CouchbaseBucketGoCB) Incr(k string, amt, def uint64, exp uint32) (
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("Incr with key: %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("Incr with key", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "Error during Incr with key: %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 6]
```diff
@@ -1352,8 +1372,10 @@ func (bucket *CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv in
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("GetWithXattr %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("GetWithXattr", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "GetWithXattr %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 7]
```diff
@@ -1278,8 +1296,10 @@ func (bucket *CouchbaseBucketGoCB) UpdateXattr(k string, xattrKey string, exp ui
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("UpdateXattr with key %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("UpdateXattr", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "Error during UpdateXattr with key %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 8]
```diff
@@ -1214,8 +1230,10 @@ func (bucket *CouchbaseBucketGoCB) WriteCasWithXattr(k string, xattrKey string,
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("WriteCasWithXattr with key %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("WriteCasWithXattr", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "WriteCasWithXattr with key %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 9]
```diff
@@ -1117,8 +1131,10 @@ func (bucket *CouchbaseBucketGoCB) WriteCas(k string, flags int, exp uint32, cas
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("WriteCas with key %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("WriteCas", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "WriteCase with key %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 10]
```diff
@@ -903,8 +915,10 @@ func (bucket *CouchbaseBucketGoCB) Touch(k string, exp uint32) (cas uint64, err
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("Touch for key %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("Touch", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, "Error during Touch for key %v", k)
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 11]
```diff
@@ -865,8 +875,10 @@ func (bucket *CouchbaseBucketGoCB) GetAndTouchRaw(k string, exp uint32) (rv []by
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("GetAndTouchRaw with key %v", k)
-	err, result := RetryLoop(description, worker, bucket.spec.RetrySleeper())
+	err, result := RetryLoop("GetAndTouchRaw", worker, bucket.spec.RetrySleeper())
+	if err != nil {
+		err = pkgerrors.Wrapf(err, fmt.Sprintf("Error during GetAndTouchRaw with key %v", k))
```
Comment (bbrks) [inline base/bucket_gocb.go]: key redaction

File: base/bucket_gocb.go [hunk 12]
```diff
@@ -2395,10 +2424,9 @@ func (bucket *CouchbaseBucketGoCB) Flush() error {
 	}
 
 	// Kick off retry loop
-	description := fmt.Sprintf("Wait until bucket %s has 0 items after flush", bucket.spec.BucketName)
-	err, _ = RetryLoop(description, worker, CreateMaxDoublingSleeperFunc(25, 100, 10000))
+	err, _ = RetryLoop("Wait until bucket has 0 items after flush", worker, CreateMaxDoublingSleeperFunc(25, 100, 10000))
 	if err != nil {
-		return err
+		return pkgerrors.Wrapf(err, "Error during Wait until bucket %s has 0 items after flush", bucket.spec.BucketName)
```
Comment (bbrks) [inline base/bucket_gocb.go]: metadata redaction on bucket name

**Review by bbrks — CHANGES_REQUESTED** [review]

File: base/bucket_gocb.go (hunk 1 above)
Comment (JRascagneres) [reply base/bucket_gocb.go]: I wanted to ensure the same behaviour was kept as some callers are checking for error types. I'll double check. Although it looks like `IsDocNotFoundError` check should handle wrapped errors

File: base/bucket_gocb.go (hunk 2 above)
Comment (JRascagneres) [reply base/bucket_gocb.go]: Same as other one

File: base/bucket_gocb.go (hunk 1 above)
Comment (bbrks) [reply base/bucket_gocb.go]: Fair enough - didn't even consider consumers of the error, but I agree it makes sense to have stuff further up the stack unwrap if nessesary and drop special handling here.

Comment (JRascagneres) [reply base/bucket_gocb.go]: Same with `IsKeyNotFoundError`

**Comment by adamcfraser** [comment]
@JRascagneres Changes look good, assigning back to you for cleanup of the failing integration tests.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**